py-polars/src/conversion.rs

```diff
@@ -632,6 +629,29 @@ impl<'s> FromPyObject<'s> for Wrap<AnyValue<'s>> {
         } else if let Ok(v) = ob.extract::<&'s [u8]>() {
             Ok(AnyValue::Binary(v).into())
         } else {
+            let mut type_name = ob.get_type().name()?;
+            if let Some(Ok(v)) = (type_name != "Decimal").then_some(ob.extract::<f64>()) {
+                return Ok(AnyValue::Float64(v).into());
+            };
+            // Can't use pyo3::types::PyDateTime with abi3-py37 feature,
+            // so need this workaround instead of `isinstance(ob, datetime)`.
+            if !["datetime", "date", "timedelta", "time"].contains(&type_name) {
```

Comment:
This is all checked in the `match` arm below, so this is quite redundant and inserts a rather expensive loop in any temporal conversion. I think we can put all this logic under the catch all `_` branch in the `match` statement below on line 754.